WalletWasabi/WabiSabi/Client/WabiSabiApiClient.cs

```diff
@@ -0,0 +1,63 @@
+using NBitcoin;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using WalletWasabi.Crypto;
+using WalletWasabi.WabiSabi.Backend.PostRequests;
+using WalletWasabi.WabiSabi.Crypto;
+using WalletWasabi.WabiSabi.Models;
+
+namespace WalletWasabi.WabiSabi.Client
+{
+	public class WabiSabiApiClient
+	{
+		public WabiSabiApiClient(WabiSabiClient wabiSabiClientAmount, WabiSabiClient wabiSabiClientWeight, IRequestHandler requestHandler)
+		{
+			WabiSabiClientAmount = wabiSabiClientAmount;
+			WabiSabiClientWeight = wabiSabiClientWeight;
+			RequestHandler = requestHandler;
+		}
+
+		public WabiSabiClient WabiSabiClientAmount { get; }
+		public WabiSabiClient WabiSabiClientWeight { get; }
+		public IRequestHandler RequestHandler { get; }
+
+		public Task RegisterInputAsync(Money amount, OutPoint outPoint, Key key, Guid roundId, uint256 roundHash) =>
+			RegisterInputAsync(
+				new[] { amount },
+				new[] { outPoint },
+				new[] { key },
+				roundId,
+				roundHash);
+				
+		public async Task RegisterInputAsync(
+			IEnumerable<Money> amounts, 
+			IEnumerable<OutPoint> outPoints, 
+			IEnumerable<Key> keys,
+			Guid roundId,
+			uint256 roundHash)
+		{
+			static byte[] GenerateOwnershipProof(Key key, uint256 roundHash) => OwnershipProof.GenerateCoinJoinInputProof(
+				key, 
+				new CoinJoinInputCommitmentData("CoinJoinCoordinatorIdentifier", roundHash)).ToBytes();
+
+			var registrableInputs = outPoints
+				.Zip(keys, (outPoint, key) => (outPoint, key))
+				.Select(x => new InputRoundSignaturePair(x.outPoint, GenerateOwnershipProof(x.key, roundHash)));
+
+			var (zeroAmountCredentialRequest, zeroAmountCredentialResponseValidation) = WabiSabiClientAmount.CreateRequestForZeroAmount();
+			var (zeroWeightCredentialRequest, zeroWeightCredentialResponseValidation) = WabiSabiClientWeight.CreateRequestForZeroAmount();
+
+			var inputRegistrationResponse = await RequestHandler.RegisterInputAsync(
+				new InputsRegistrationRequest(
+					roundId,
+					registrableInputs,
+					zeroAmountCredentialRequest,
+					zeroWeightCredentialRequest));
+
+			WabiSabiClientAmount.HandleResponse(inputRegistrationResponse.AmountCredentials, zeroAmountCredentialResponseValidation);
+			WabiSabiClientWeight.HandleResponse(inputRegistrationResponse.WeightCredentials, zeroWeightCredentialResponseValidation);
```

Comment:
Any ideas about what to make the high level API do with the results of this?

Eventually this client will need to:

1. track the Alice ID
2. make 1 or more confirmation requests with both real and a zero credential request each in a loop
3. make sure that only a tor circuit is used with at most one Alice ID (but multiple tor circuits could be used for a single Alice ID in case of failure)
4. send the final signature when done.

At this stage i think this only means making sure the ID is preserved, especially if the high level API client takes care of the connection confirmation loop.

What I'm imagining is that the higher level consumers of this class would basically provide some UTXOs to register, and a "plan" for output registrations to make.

Like the confirmation loop, I'm still not sure whether things like credential merging/splitting (the `k` parameter constrains the precise seqeuence) and issues like timing, or changing the plan (the latter will become very important when considering payments from CoinJoins) should be in scope at this level of abstraction, or one level above.

Reply:
Let me clarify the design goals behind this code. This `WabiSabiApiClient` (ugly name) is a mirror of the server-side `PostRequestHandler`, this means that it contains the same methods: `RegisterInputAsync`, `RegisterOutputAsync`, `SignTransactionAsync`, etc. It is like a proxy pattern. The idea is to have in the client-side a simple way to get access to the functionality provided by the server.

My idea is that other clients, AliceClient & BobClient, use a `WabiSabiApiClient` instance to communicate with the coordinator abstracting away the complexity of creating credentials for amounts an weight, proving ownership of the inputs and things like that.

I think the `RegisterInputAsync` method will be invoked by the AliceClient and it has to return the `AliceID` so the `AliceClient` can keep track of if. I mean, this client should be just a proxy and be quite dumb. The same applies for the loop of connection confirmations, that loop should be part of the AliceClient client.

Another important detail of this design is that it communicates with an `IRequestHandler`, currently the only class that implements that interface is the real server-side `PostRequestHandler` what allows us to test end-to-end interaction. The same abstraction should be used in the future to implement a class that takes care to serialize the messages and send/receive using http protocol and make sure to switch Tor identities and all the things involving communication. In other words, the `IRequestHandler` is to decouple the server and to abstract away the communication details.

Reply:
ACK, and I especially agree with the approach of making interface types like that, in my experience that is the cleanest way to build protocol implementations (although not in C#).

So in that case everything i mentioned should out of scope, and this code should be "for real", not the throwaway stepping stone we discussed. On top of this we can build the higher level clients for the specific roles, starting with throwaway versions just for testing purposes.

Unfortunately though, another layer of abstraction means another naming problem to deal with ;-)